[`core`/ FEAT] Add the possibility to push custom tags using `PreTrainedModel` itself

younesbelkada opened the pull request:
# What does this PR do?

From an idea we discussed internally. This PR introduces a new API to inject custom tags in the model card

From a community perspective it will make easier to push custom tags as for now it is only limited to trainers. 

Below I demonstrate the simplicity of the API

```python
from transformers import AutoModelForCausalLM

model_name = "HuggingFaceM4/tiny-random-LlamaForCausalLM"
model = AutoModelForCausalLM.from_pretrained(model_name)

model.add_model_tags(["tag-test"])
model.push_to_hub("llama-tagged")
```

cc @osanseviero @Narsil @julien-c 

Note with the current design each time a user calls `push_to_hub`, it will create a model card template if no model card is present on the hub

younesbelkada committed:
v1 tags

younesbelkada committed:
remove unneeded conversion

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28405). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

younesbelkada committed:
v2

younesbelkada commented:
I made some modification after some offline discussion, an exmaple model pushed can be found here: https://huggingface.co/ybelkada/test-tags-model lmk wdyt @amyeroberts @ArthurZucker @Wauplin

younesbelkada committed:
rm unneeded warning

amyeroberts commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
+        card_data = ModelCardData(language="en", tags=[])
+        model_card = ModelCard.from_template(card_data)
+
+    if model_card is not None and tags is not None:
```
Why would `model_card` be `None` here?

amyeroberts commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
     return {"repo": cache_repo, "revision": revision, "filename": filename}
 
 
+def create_and_tag_model_card(repo_id, tags=None):
```
Let's include the types so someone doesn't have to look into the internals to know what to pass

```suggestion
def create_and_tag_model_card(repo_id: str, tags: Optional[List[str]] = None):
```

Wauplin replied:
+1 on static typing :)

Also maybe naming the method `generate_model_card` would be more appropriate as it implies creating + adding metadata to it (and not only tags)

amyeroberts commented on src/transformers/modeling_utils.py:
```diff
@@ -1239,6 +1243,24 @@ def _backward_compatibility_gradient_checkpointing(self):
+        if self._model_tags is None:
+            self._model_tags = []
+
+        for tag in tags:
+            if tag not in self._model_tags:
+                self._model_tags.append(tag)
```
I don't think we want an append behaviour here - more of an overwrite. Otherwise, if I accidentally add a tag there's no way to ever remove it. I'd suggest a `set_model_tags` and `add_model_tags` method

Wauplin replied:
I'd also be in favor of switching to `add_model_tags` if we are appending tags to a list.
Don't know if adding support for removing tags if really necessary (but I don't have full context).

younesbelkada replied:
Yeah makes sense, I changed that method to `add_tags` and added a new method `set_tags`

amyeroberts reviewed (commented):
Thanks for adding this! Looking good 🤗 

Two comments: 
* I'd have some way for the user to inspect the current tags e.g. `get_model_tags` or just using a non-private attribute `model_tags`. Otherwise I'm blindly modifying the tags without knowing the current state. 
* It should be possible to reset or remove tags, rather than just add them

Wauplin commented on src/transformers/modeling_utils.py:
```diff
@@ -2472,6 +2471,16 @@ def save_pretrained(
+        elif "tags" in kwargs and self._model_tags is not None:
+            logger.warning(
+                "You manually passed `tags` to `push_to_hub` method and the model has already some tags set, we will use the tags that you passed."
+            )
```
+1 on merging both lists if they exist (for flexibility)

younesbelkada replied:
Makes sense, done !

Wauplin commented on src/transformers/modeling_utils.py:
```diff
@@ -2433,6 +2466,16 @@ def save_pretrained(
+        if "tags" not in kwargs:
+            kwargs["tags"] = self._model_tags
+        elif "tags" in kwargs and self._model_tags is not None:
+            logger.warning(
```
(nit) I think using `warnings.warn` would be more appropriate here. I usually tend to follow the rule from https://docs.python.org/3/howto/logging.html:

> [warnings.warn()](https://docs.python.org/3/library/warnings.html#warnings.warn) in library code if the issue is avoidable and the client application should be modified to eliminate the warning

> [logging.warning()](https://docs.python.org/3/library/logging.html#logging.warning) if there is nothing the client application can do about the situation, but the event should still be noted

Here the problem is most likely avoidable by the user.

younesbelkada replied:
Thanks! I think that transformers fully switched to using `logger.warning` (I got told that in a PR review from @LysandreJik but I don't remember which one 😢 ) let me try to find that and link the comment here

LysandreJik replied:
No you can keep using `warnings.warn`! see https://github.com/huggingface/transformers/pull/26527

Wauplin commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
+    """
+    Creates a dummy model card and tags it.
+    """
+    model_card = None
+
```
```suggestion
```
(is it necessary?)

younesbelkada replied:
True yeah this is not necessary indeed !

Wauplin commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
+        model_card = ModelCard.load(repo_id)
+    except EntryNotFoundError:
+        # Otherwise create a simple model card from template
+        card_data = ModelCardData(language="en", tags=[])
```
(not sure) is it expected that all models should be tagged as an "English" model?

Also could we add `library_name` and `model_name` as well?

younesbelkada replied:
Done ! I also removed the default `en` language tag

Wauplin commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
+    except EntryNotFoundError:
+        # Otherwise create a simple model card from template
+        card_data = ModelCardData(language="en", tags=[])
+        model_card = ModelCard.from_template(card_data)
```
would be nice to set a default `model_summary` field here with some info along the lines "this is a *** model. It has been pushed to the Hub using `transformers`. This model card has been auto-generated."

younesbelkada replied:
Makes sense, done !

Wauplin commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1093,27 @@ def extract_info_from_url(url):
+        model_card = ModelCard.from_template(card_data)
+
+    if model_card is not None and tags is not None:
+        for model_tag in tags:
+            model_card.data.tags.append(model_tag)
```
```suggestion
        for model_tag in tags:
            if model_tag not in model_card.data.tags:
                model_card.data.tags.append(model_tag)
```
(just in case tag already exists on the Hub)

Wauplin reviewed (commented):
Thanks for implementing this @younesbelkada! 
I left some comments. Some might be out of scope for this PR so feel free to do as you prefer :)

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into set-custom-tag

younesbelkada committed:
add more utility methods

younesbelkada committed:
Update src/transformers/utils/hub.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/utils/hub.py

Co-authored-by: Lucain <lucainp@gmail.com>

younesbelkada committed:
Update src/transformers/utils/hub.py

Co-authored-by: Lucain <lucainp@gmail.com>

younesbelkada committed:
more enhancements

younesbelkada committed:
oops

younesbelkada committed:
merge tags

younesbelkada committed:
clean up

younesbelkada committed:
revert unneeded change

younesbelkada commented:
@amyeroberts @Wauplin I think that I have addressed most of your comments now, the API is much more flexible and less "agressive" thanks to your suggestions
1- Creates a simple description
2- Moved `transformers` to `library-name`
3- Now Trainer and `model_tags` are compatible together, which was not the case in the first commits. e.g. when combining `SFTTrainer` from trl that automatically pushes `sft` and `trl` tags and this PR, all tags gets successfully pushed. E.g.:
```python
from transformers import AutoModelForCausalLM, TrainingArguments
from trl import SFTTrainer
from datasets import load_dataset

model_name = "HuggingFaceM4/tiny-random-LlamaForCausalLM"
model = AutoModelForCausalLM.from_pretrained(model_name)

model.add_model_tags(["tag-test"])

train_data = load_dataset("imdb", split="train")

trainer = SFTTrainer(
    model=model,
    args=TrainingArguments(output_dir="test-tag-already-tagged"),
    train_dataset=train_data,
    dataset_text_field="text"
)

trainer.push_to_hub("ybelkada/test-tag-already-tagged")
```
The example repo is here: 
4- If tags already exists it will not overwrite all the tags but add the new tags with existing tags. e.g. if one does `model.add_tags(["test-tag-2"])` after already pushing a model with a tag to Hub, as you can see from this repository: https://huggingface.co/ybelkada/test-model-already-tagged the new tag gets appended

cc @osanseviero as well

Wauplin reviewed (commented):
Looks good! Thanks for making the changes :)

Regarding `set_model_tags`, `reset_model_tags` and `remove_model_tags`, I feel that their behavior might be misleading/non-explicit. For example `remove_model_tags` will remove tags from the model object (`AutoModelForCausalLM.model_tags` in your example) but not from the model card on the Hub. I think some users can expect `model.remove_model_tags("sft")` to remove the `"sft"` tag from the existing model card. Maybe it's safer not to implement those three methods to avoid questions. (Also I'm not sure they would be really used -only my 2c-)

amyeroberts reviewed (approved):
Thanks for iterating - LGTM! 

Agree with @Wauplin re other methods. I realise I asked for `set_model_tags` but realise wow `model_tags` are public, we probably don't need it as the user can just do it themselves directly

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into set-custom-tag

younesbelkada committed:
add extensive docs

younesbelkada committed:
more docs

younesbelkada committed:
more kwargs

younesbelkada commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1095,43 @@ def extract_info_from_url(url):
+    repo_id: str,
+    tags: Optional[List[str]] = None,
+    token: Optional[str] = None,
+    ignore_metadata_errors: bool = False,
```
@Wauplin @amyeroberts fine to assume that `repo_type` will always be set to `"model"` in `ModelCard.load()` (inspecting at that methods signature)

Wauplin replied:
Yes definitely, `repo_type="model"` is implicit when using `ModelCard`.
(the parameter exists because `ModelCard` inherits from the generic `RepoCard`. Not a perfect design-choice but it's fine :) )

younesbelkada replied:
ok perfect! let's keep it like that then

younesbelkada reviewed (commented):
Thanks @amyeroberts @Wauplin , this is ready for a final review, I tested it on different scenarios and added some docs with a simple code snippet, let me know wdyt !

younesbelkada committed:
add test

younesbelkada committed:
oops

younesbelkada committed:
fix test

younesbelkada commented on src/transformers/utils/hub.py:
```diff
@@ -1081,6 +1095,43 @@ def extract_info_from_url(url):
+            The repo_id where to look for the model card.
+        tags (`List[str]`, *optional*):
+            The list of tags to add in the model card
+        token (`str`, *optional*):
+            Authentication token, obtained with `huggingface_hub.HfApi.login` method. Will default to the stored token.
+        ignore_metadata_errors (`str`):
+            If True, errors while parsing the metadata section will be ignored. Some information might be lost during
+            the process. Use it at your own risk.
```
I copied this from ModelCard.load_card docstring

Wauplin commented on src/transformers/modeling_utils.py:
```diff
@@ -1253,6 +1255,38 @@ def _backward_compatibility_gradient_checkpointing(self):
 
+    def add_model_tags(self, tags: Union[List[str], str]) -> None:
+        r"""
+        Add custom tags into the model that gets pushed in 🤗 Hub. Will
```
(nit)
```suggestion
        Add custom tags into the model that gets pushed to the Hugging Face Hub. Will
```

younesbelkada replied:
let me push your fix as well

Wauplin replied:
Yep sorry I was  already reviewing when Omar posted his comment :)

osanseviero commented on src/transformers/modeling_utils.py:
```diff
@@ -1253,6 +1255,38 @@ def _backward_compatibility_gradient_checkpointing(self):
 
+    def add_model_tags(self, tags: Union[List[str], str]) -> None:
+        r"""
+        Add custom tags into the model that gets pushed in 🤗 Hub. Will
```
```suggestion
        Add custom tags into the model that gets pushed to the Hub. Will
```

Wauplin commented on src/transformers/utils/hub.py:
```diff
@@ -762,6 +764,7 @@ def push_to_hub(
         safe_serialization: bool = True,
         revision: str = None,
         commit_description: str = None,
+        tags: List[str] = None,
```
```suggestion
        tags: Optional[List[str]] = None,
```

`Optional` since default value is None?

osanseviero reviewed (approved):
Looks cool!

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Omar Sanseviero <osanseviero@gmail.com>

Wauplin commented on tests/test_modeling_utils.py:
```diff
@@ -1515,6 +1520,28 @@ def test_push_to_hub_dynamic_model(self):
+
+        self.assertTrue(model.model_tags == new_tags)
+
+        model.push_to_hub("test-dynamic-model-with-tags", token=self._token)
```
Isn't it problematic to have a fixed name for the repo_id? This could lead to conflicts if several CIs are ran in parallel. Using a unique id (uuid) would avoid that.
(not an expert of `transformers`'s CI though. It's just that I've experienced similar problem in `huggingface_hub`'s CI in the past)

younesbelkada replied:
Hmm good point, we do the same things for all tests above so perhaps it is fine, I will let @ydshieh comment on this if he has more insights than me

Wauplin reviewed (approved):
Looks good to me! I left some minor comments but feel free to ignore.

younesbelkada committed:
Update src/transformers/utils/hub.py

Co-authored-by: Lucain <lucainp@gmail.com>

younesbelkada committed:
Update src/transformers/modeling_utils.py

amyeroberts reviewed (commented):
Thanks for iterating! 

Just a final comment / question on the push_to_hub logic

younesbelkada committed:
Update src/transformers/trainer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

LysandreJik commented:
Thanks for handling this @younesbelkada

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2447,6 +2490,17 @@ def save_pretrained(
 
+    @wraps(PushToHubMixin.push_to_hub)
+    def push_to_hub(self, *args, **kwargs):
+        if "tags" not in kwargs:
+            kwargs["tags"] = self.model_tags
+        elif "tags" in kwargs and self.model_tags is not None:
+            for model_tag in self.model_tags:
+                # merge the tags together
+                if model_tag not in kwargs["tags"]:
+                    kwargs["tags"].append(model_tag)
+        return super().push_to_hub(*args, **kwargs)
```
I think your suggestion sounds great, also 
> Is it possible kwargs["tags"] is a string here too?
It can be strings, let me adapt a bit the logic after accepting your suggestion

younesbelkada replied:
I just took care of the str case in [1e3fc1e](https://github.com/huggingface/transformers/pull/28405/commits/1e3fc1e306bbf953fe513246273fe2544f057288)

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into set-custom-tag

younesbelkada committed:
add more conditions

younesbelkada commented on src/transformers/trainer.py:
```diff
@@ -3581,6 +3581,15 @@ def create_model_card(
             is_peft_library = library_name == "peft"
 
+            # Append existing tags in `tags`
+            existing_tags = ModelCard.load(model_card_filepath).data.tags
```
This handles a corner case I just found out for Trainers, if you push into a repo that has already some tags, it will overwrite the existing tags with new tags. This block simply cirumvents this by instead of overwriting new tags it will append new tags from existing tags

amyeroberts replied:
Look OK to me - only comment is that `existing_tags` might be `None`, so you'll need to convert it to a list before adding `tags` if `ModelCard.load(model_card_filepath).data.tags` doesn't have anything set

younesbelkada replied:
I pushed in 59738c67c273e7a86b13f6648446fff21367151d a simple workaround, I think that way we don't need to convert it to a list as we care about that corner case only if `tags` are explicilty passed + there are some existing tags already

younesbelkada committed:
more logic

amyeroberts reviewed (approved):
Thanks for adding this and iterating - looks great!
